interfaces/builtin/gpio_chardev.go

```diff
@@ -0,0 +1,212 @@
+	}
+	// Check that range is not unrealistically large.
+	if r.Size() > maxLinesCount {
+		return fmt.Errorf(`invalid "lines" attribute: range size cannot exceed %d, found %d`, maxLinesCount, r.Size())
+	}
+
+	return nil
+}
+
+func (iface *gpioChardevInterface) BeforeConnectPlug(plug *interfaces.ConnectedPlug) error {
+	// gpio-chardev is hidden behind an experimental feature flag until kernel
+	// improvements for the gpio-aggregator interface lands.
+	// https://lore.kernel.org/all/20250203031213.399914-1-koichiro.den@canonical.com
+	if !features.GPIOChardevInterface.IsEnabled() {
+		_, flag := features.GPIOChardevInterface.ConfigOption()
+		return fmt.Errorf("gpio-chardev interface requires the %q flag to be set", flag)
+	}
+	return nil
+}
+
+func (iface *gpioChardevInterface) SystemdConnectedSlot(spec *systemd.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var sourceChip []string
+	if err := slot.Attr("source-chip", &sourceChip); err != nil {
+		return err
+	}
+	var lines string
+	if err := slot.Attr("lines", &lines); err != nil {
+		return err
+	}
+	snapName := slot.Snap().InstanceName()
+	slotName := slot.Name()
+
+	serviceSuffix := fmt.Sprintf("gpio-chardev-%s", slotName)
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		// snap-gpio-helper export-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStart: fmt.Sprintf("%s/snap-gpio-helper export-chardev %q %q %q %q",
+			dirs.DistroLibExecDir, strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snap-gpio-helper unexport-chardev "<chip-labels>" "<lines>" "<gadget>" "<slot-name>"
+		ExecStop: fmt.Sprintf("%s/snap-gpio-helper unexport-chardev %q %q %q %q",
+			dirs.DistroLibExecDir, strings.Join(sourceChip, ","), lines, snapName, slotName),
+		// snapd.gpio-chardev-setup.target is used for synchronization of
+		// app services and virtual device setup during boot.
+		WantedBy: "snapd.gpio-chardev-setup.target",
+		Before:   "snapd.gpio-chardev-setup.target",
+	}
+	return spec.AddService(serviceSuffix, service)
+}
+
+func (iface *gpioChardevInterface) SystemdConnectedPlug(spec *systemd.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	slotName := slot.Name()
+	slotSnapName := slot.Snap().InstanceName()
+	plugName := slot.Name()
+	plugSnapName := plug.Snap().InstanceName()
+
+	target := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", slotSnapName, slotName)
+	symlink := fmt.Sprintf("/dev/snap/gpio-chardev/%s/%s", plugSnapName, plugName)
+
+	// Create symlink pointing to exported virtual slot device.
+	serviceSuffix := fmt.Sprintf("gpio-chardev-%s", plugName)
+	service := &systemd.Service{
+		Type:            "oneshot",
+		RemainAfterExit: true,
+		ExecStart:       fmt.Sprintf("/bin/sh -c 'mkdir -p %q && ln -s %q %q'", filepath.Dir(symlink), target, symlink),
+		ExecStop:        fmt.Sprintf("/bin/sh -c 'rm -f %q'", symlink),
```

Comment:
Do we need to run this via the shell?

Reply:
not really, I was keeping it consistent with `ExecStart`, I can switch to `/bin/rm`